subsys/bluetooth/controller/ll_sw/ctrl.c

```diff
@@ -9363,21 +9373,25 @@ static void prepare_pdu_data_tx(struct connection *conn,
 				struct pdu_data **pdu_data_tx)
 {
 	struct pdu_data *_pdu_data_tx;
+	bool pause_tx = false;
 
 	if (/* empty packet */
 	    conn->empty ||
 	    /* no ctrl or data packet */
 	    !conn->pkt_tx_head ||
 	    /* data tx paused, only control packets allowed */
 	    ((
-#if defined(CONFIG_BT_CTLR_LE_ENC)
-	      conn->pause_tx ||
-	      is_enc_req_pause_tx(conn) ||
-#endif /* CONFIG_BT_CTLR_LE_ENC */
 #if defined(CONFIG_BT_CTLR_DATA_LENGTH)
 	      conn->llcp_length.pause_tx ||
 #endif /* CONFIG_BT_CTLR_DATA_LENGTH */
-	      0) && (conn->pkt_tx_head != conn->pkt_tx_ctrl))) {
+#if defined(CONFIG_BT_CTLR_LE_ENC)
+	      conn->pause_tx ||
+	      /* Encryption setup queued */
+	      (pause_tx = is_enc_req_pause_tx(conn)) ||
+#endif /* CONFIG_BT_CTLR_LE_ENC */
+	      0) &&
+	    /* Encryption setup queued or data paused */
+	    (pause_tx || (conn->pkt_tx_head != conn->pkt_tx_ctrl)))) {
```

Comment:
+1 for comments, that helps a lot.
But it is not obvious if all the corner cases of this massive-if are properly handled.

For complex branching like this, I personally like to factor out the logical meaning across const booleans, and let the booleans derive from the previous ones. With those, I can name them and comment on them to attain more readability.

Reply:
I agree this is complex (have been on my mind, and gave it almost a week before I pushed ;-) ). It is this way trying to short circuit as early as possible to reduce CPU usage. But I may be incorrect based on how the compiler/linker could optimize if the code was written in a more readable fashion.

Example, a connection with only empty) packets use less CPU in comparison to when data is in the queue. And during a active encryption procedure, the call to is_enc_req_pause_tx is completely bypassed by conn->pause_tx, etc.

I am open to alternative implementations that meet the constraints.